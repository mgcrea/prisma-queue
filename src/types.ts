import type { Prisma, PrismaClient, QueueJob as PrismaQueueJob } from "@prisma/client";
import type { PrismaJob } from "./PrismaJob";

export type JobPayload = Prisma.InputJsonValue;
export type JobResult = Prisma.InputJsonValue;
export type DatabaseJob<T, U> = Omit<PrismaQueueJob, "payload" | "result"> & { payload: T; result: U };

export type JobCreator<T extends JobPayload> = (client: PrismaLightClient) => Promise<T>;
export type JobWorker<T extends JobPayload = JobPayload, U extends JobResult = JobResult> = (
  job: PrismaJob<T, U>,
  client: PrismaClient
) => Promise<U>;

export type PrismaLightClient = Omit<
  PrismaClient,
  "$connect" | "$disconnect" | "$on" | "$transaction" | "$use"
>;
