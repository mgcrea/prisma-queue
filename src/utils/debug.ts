import createDebug from 'debug';

export const debug = createDebug('prisma-queue');
