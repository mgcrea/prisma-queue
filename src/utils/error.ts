export const serializeError = (err: unknown) => {
  if (err instanceof Error) {
    return {
      name: err.name,
      message: err.message,
      stack: err.stack,
    };
  }
  return {
    name: "UnknownError",
    message: `${err}`,
    stack: undefined,
  };
};
