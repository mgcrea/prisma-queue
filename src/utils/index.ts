export * from './debug';
export * from './error';
export * from './prisma';
export * from './string';
export * from './time';
