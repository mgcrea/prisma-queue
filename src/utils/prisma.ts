import {Prisma} from '@prisma/client';
import assert from 'assert';

export const getTableName = (modelName: string): string => {
  const model = Prisma.dmmf.datamodel.models.find((model) => model.name === modelName);
  assert(model && model.dbName, `Did not foudn model=${modelName} in Prisma.dmmf!`);
  return model.dbName;
};
