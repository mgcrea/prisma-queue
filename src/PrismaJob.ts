import type { Prisma } from "@prisma/client";
import type { DatabaseJob, PrismaLightClient } from "./types";
import { isPrismaError } from "./utils";
// import { debug } from "./utils";

export type PrismaJobOptions = {
  model: Prisma.QueueJobDelegate;
  client: PrismaLightClient;
};

/**
 * Represents a job within a Prisma-managed queue.
 */
export class PrismaJob<Payload, Result> {
  #model: Prisma.QueueJobDelegate;
  #client: PrismaLightClient;
  #record: DatabaseJob<Payload, Result>;

  public readonly id;
  public readonly createdAt: Date = new Date();

  /**
   * Constructs a new PrismaJob instance with the provided job record and database access objects.
   * @param record - The initial database job record.
   * @param model - The Prisma delegate used for database operations related to the job.
   * @param client - The Prisma client for executing arbitrary queries.
   */
  constructor(record: DatabaseJob<Payload, Result>, { model, client }: PrismaJobOptions) {
    this.#model = model;
    this.#client = client;
    this.#record = record;
    this.id = record.id;
  }

  /**
   * Internal method to assign a new record to the job.
   * @param record - Optional new record to assign.
   */
  #assign(record?: DatabaseJob<Payload, Result>) {
    if (record) {
      this.#record = record;
    }
  }

  /**
   * Gets the current job record.
   */
  public get record(): DatabaseJob<Payload, Result> {
    return this.#record;
  }

  /**
   * Gets the job's unique key if any.
   */
  public get key() {
    return this.#record.key;
  }

  /**
   * Gets the job's queue name.
   */
  public get queue() {
    return this.#record.queue;
  }

  /**
   * Gets the CRON expression associated with the job for recurring scheduling.
   */
  public get cron() {
    return this.#record.cron;
  }

  /**
   * Gets the job's priority level.
   */
  public get priority() {
    return this.#record.priority;
  }

  /**
   * Gets the payload associated with the job.
   */
  public get payload() {
    return this.#record.payload;
  }

  /**
   * Gets the timestamp when the job was finished.
   */
  public get finishedAt() {
    return this.#record.finishedAt;
  }

  /**
   * Gets the error record if the job failed.
   */
  public get error() {
    return this.#record.error;
  }

  /**
   * Updates the job's progress percentage.
   * @param progress - The new progress percentage.
   */
  public async progress(progress: number): Promise<DatabaseJob<Payload, Result>> {
    return await this.update({ progress: Math.max(0, Math.min(100, progress)) });
  }

  /**
   * Fetches the latest job record from the database and updates the internal state.
   */
  public async fetch(): Promise<DatabaseJob<Payload, Result>> {
    const record = (await this.#model.findUnique({
      where: { id: this.id },
    })) as DatabaseJob<Payload, Result>;
    this.#assign(record);
    return record;
  }

  /**
   * Updates the job record in the database with new data.
   * @param data - The new data to be merged with the existing job record.
   */
  public async update(data: Prisma.QueueJobUpdateInput): Promise<DatabaseJob<Payload, Result>> {
    const record = (await this.#model.update({
      where: { id: this.id },
      data,
    })) as DatabaseJob<Payload, Result>;
    this.#assign(record);
    return record;
  }

  /**
   * Deletes the job from the database.
   */
  public async delete(): Promise<DatabaseJob<Payload, Result>> {
    const record = (await this.#model.delete({
      where: { id: this.id },
    })) as DatabaseJob<Payload, Result>;
    return record;
  }

  /**
   * Checks if the job is currently locked by another transaction.
   * @returns {Promise<boolean>} True if the job is locked, false otherwise.
   */
  public async isLocked(): Promise<boolean> {
    try {
      // Attempt to select and lock the row with a timeout
      await this.#client.$executeRawUnsafe(
        `SELECT "id" FROM "public"."queue_jobs" WHERE "id" = $1 FOR UPDATE NOWAIT`,
        this.id,
      );

      // If we reach here, the row is not locked
      return false;
    } catch (error) {
      // Handle specific error types that indicate lock contention
      if (isPrismaError(error) && error.meta?.["code"] === "55P03") {
        // PostgreSQL's lock_not_available
        return true;
      }
      // Re-throw other unexpected errors
      throw error;
    }
  }
}
