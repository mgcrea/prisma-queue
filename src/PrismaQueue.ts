import {Prisma, PrismaClient} from '@prisma/client';
import assert from 'assert';
import Cron from 'croner';
import {EventEmitter} from 'events';
import {PrismaJob} from './PrismaJob';
import {DatabaseJob, JobCreator, JobPayload, JobResult, JobWorker} from './types';
import {calculateDelay, debug, escape, getTableName, serializeError, waitFor} from './utils';

export type PrismaQueueOptions<T, U> = {
  prisma?: PrismaClient;
  name?: string;
  maxConcurrency?: number;
  pollInterval?: number;
  // jobInterval?: number;
  tableName?: string;
};

export type EnqueueOptions = {
  runAt?: Date;
  key?: string;
  maxAttempts?: number;
  priority?: number;
};
export type ScheduleOptions = Omit<EnqueueOptions, 'key' | 'runAt'> & {
  key: string;
  cron: string;
};

const JOB_INTERVAL = 25;

export class PrismaQueue<T extends JobPayload = JobPayload, U extends JobResult = JobResult> extends EventEmitter {
  private name: string;
  private prisma: PrismaClient;
  private config: Required<Omit<PrismaQueueOptions<T, U>, 'name' | 'prisma'>>;

  private count = 0;
  private concurrency = 0;
  private stopped = true;

  public constructor(private options: PrismaQueueOptions<T, U> = {}, public worker?: JobWorker<T, U>) {
    super();

    const {
      prisma = new PrismaClient(),
      name = 'default',
      tableName = getTableName('QueueJob'),
      maxConcurrency = 1,
      pollInterval = 10000,
      // jobInterval = 100,
    } = options;

    assert(name.length <= 255, 'name must be less or equal to 255 chars');
    assert(pollInterval >= 100, 'pollInterval must be more than 100 ms');
    // assert(jobInterval >= 10, 'jobInterval must be more than 10 ms');

    this.name = name;
    this.prisma = prisma;
    this.config = {
      tableName,
      maxConcurrency,
      pollInterval,
      // jobInterval,
    };
  }

  private get model() {
    return this.prisma.queueJob;
  }

  public async start(): Promise<void> {
    this.stopped = false;
    return this.poll();
  }

  public async stop(): Promise<void> {
    this.stopped = true;
  }

  public async enqueue(payloadOrFunction: T | JobCreator<T>, options: EnqueueOptions = {}): Promise<PrismaJob<T, U>> {
    debug(`enqueue(${JSON.stringify(payloadOrFunction)})`);
    const {prisma, name: queueName} = this;
    const {key, maxAttempts, priority = 0, runAt} = options;
    const record = await prisma.$transaction(async (client) => {
      const payload = payloadOrFunction instanceof Function ? await payloadOrFunction(client) : payloadOrFunction;
      const data = {queue: queueName, payload, maxAttempts, priority, runAt};
      if (key && runAt) {
        const {count} = await this.model.deleteMany({
          where: {
            key,
            runAt: {
              gte: new Date(),
              not: runAt,
            },
          },
        });
        if (count > 0) {
          debug(`deleted ${count} conflicting upcoming queue jobs`);
        }
        return await this.model.upsert({
          where: {key_runAt: {key, runAt}},
          create: {key, ...data},
          update: data,
        });
      }
      return await this.model.create({data});
    });
    return new PrismaJob(record as DatabaseJob<T, U>, {prisma});
  }

  public async schedule(options: ScheduleOptions, payloadOrFunction: T | JobCreator<T>): Promise<PrismaJob<T, U>> {
    debug(`schedule(${JSON.stringify(options)})`);
    const {key, cron, ...otherOptions} = options;
    const runAt = Cron(cron).next();
    assert(runAt, `Failed to find a future occurence for given cron`);
    return this.enqueue(payloadOrFunction, {key, runAt, ...otherOptions});
  }

  private async poll(): Promise<void> {
    if (this.stopped) {
      return;
    }
    const {maxConcurrency, pollInterval} = this.config;
    let estimatedQueueSize = await this.size();
    while (estimatedQueueSize > 0) {
      while (estimatedQueueSize > 0 && this.concurrency < maxConcurrency) {
        estimatedQueueSize--;
        this.concurrency++;
        setImmediate(() =>
          this.dequeue()
            .then((count) => {
              if (count === 0) {
                estimatedQueueSize = 0;
              }
            })
            .catch((err) => this.emit('error', err))
            .finally(() => this.concurrency--)
        );
      }

      await waitFor(JOB_INTERVAL);
    }

    debug(`poll waiting for pollInterval=${pollInterval}`);
    await waitFor(pollInterval);

    await this.poll();
  }

  // https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/
  // @TODO https://docs.quirrel.dev/api/queue
  // https://github.com/quirrel-dev/quirrel/blob/main/package.json
  private async dequeue(): Promise<number | undefined> {
    debug(`dequeue()`);
    if (this.stopped) {
      return;
    }
    const {prisma, name: queueName} = this;
    const tableName = escape(this.config.tableName);
    return await prisma.$transaction(
      async (client) => {
        const rows = await client.$queryRawUnsafe<DatabaseJob<T, U>[]>(
          `UPDATE ${tableName} SET "processedAt" = NOW(), "attempts" = "attempts" + 1
           WHERE id = (
             SELECT id
             FROM ${tableName}
             WHERE (${tableName}."queue" = $1)
               AND (${tableName}."runAt" < NOW())
               AND (${tableName}."finishedAt" IS NULL)
               AND (${tableName}."notBefore" IS NULL OR ${tableName}."notBefore" < NOW())
             ORDER BY ${tableName}."priority" ASC, ${tableName}."runAt" ASC
             FOR UPDATE SKIP LOCKED
             LIMIT 1
           )
           RETURNING *;`,
          queueName
        );
        // d({ rows });
        if (rows.length < 1) {
          return 0;
        }
        const {id, payload, attempts} = rows[0];
        const job = new PrismaJob(rows[0], {prisma: client});
        let result;
        try {
          assert(this.worker, 'Missing queue worker to process job');
          debug(`starting worker for job({id: ${id}, payload: ${JSON.stringify(payload)}})`);
          result = await this.worker(job, prisma);
          await client.queueJob.update({
            where: {id: id},
            data: {finishedAt: new Date(), progress: 100, result, error: Prisma.DbNull},
          });
        } catch (err) {
          debug(`failed finishing job({id: ${id}, payload: ${JSON.stringify(payload)}}) with error="${err}"`);
          const delay = calculateDelay(attempts);
          const nextRunAt = new Date(Date.now() + delay);
          debug(`will retry at runAt=${nextRunAt.toISOString()} (attempts=${attempts})`);
          await client.queueJob.update({
            where: {id: id},
            data: {
              failedAt: new Date(),
              error: serializeError(err),
              notBefore: new Date(Date.now() + calculateDelay(attempts)),
            },
          });
        }
        return rows.length;
      },
      // @NOTE https://github.com/prisma/prisma/issues/11565#issuecomment-1031380271
      {timeout: 864e5}
    );
  }

  public async size(): Promise<number> {
    const {prisma, name: queueName} = this;
    return await prisma.queueJob.count({where: {queue: queueName, finishedAt: null}});
  }
}
