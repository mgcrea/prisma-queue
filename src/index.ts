import {PrismaQueueOptions, PrismaQueue} from './PrismaQueue';
import {JobPayload, JobResult, JobWorker} from './types';

export * from './PrismaQueue';
export * from './PrismaJob';

export const createQueue = <T extends JobPayload = JobPayload, U extends JobResult = JobResult>(
  options: PrismaQueueOptions,
  worker?: JobWorker<T, U>
) => {
  return new PrismaQueue<T, U>(options, worker);
};
