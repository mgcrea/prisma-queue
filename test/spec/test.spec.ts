import {prisma, waitFor} from 'test/utils';
import {createQueue} from 'src/index';

type JobPayload = {email: string};
type JobResult = {code: string};

// beforeAll(() => {});
describe('PrismaQueue', () => {
  it('should properly create a new queue', () => {
    const emailQueue = createQueue<JobPayload, JobResult>(
      {prisma, name: 'email', pollInterval: 2000},
      async ({email}, _db) => {
        console.log(`Sending email=${email}...`);
        await waitFor(100);
        if (email === '2') {
          throw new Error('dafuck!');
        }
        await waitFor(3000);
        console.log(`Sent email=${email}!`);
        return {code: '200'};
      }
    );
    expect(emailQueue).toBeDefined();
    expect(Object.keys(emailQueue)).toEqual('foo');
  });
});

/*

beforeAll(async () => {
  // create product categories
  await prisma.category.createMany({
    data: [{name: 'Wand'}, {name: 'Broomstick'}],
  });

  console.log('✨ 2 categories successfully created!');

  // create products
  await prisma.product.createMany({
    data: [
      {
        name: 'Holly, 11", phoenix feather',
        description: 'Harry Potters wand',
        price: 100,
        sku: 1,
        categoryId: 1,
      },
      {
        name: 'Nimbus 2000',
        description: 'Harry Potters broom',
        price: 500,
        sku: 2,
        categoryId: 2,
      },
    ],
  });

  console.log('✨ 2 products successfully created!');

  // create the customer
  await prisma.customer.create({
    data: {
      name: 'Harry Potter',
      email: 'harry@hogwarts.io',
      address: '4 Privet Drive',
    },
  });

  console.log('✨ 1 customer successfully created!');
});

afterAll(async () => {
  const deleteOrderDetails = prisma.orderDetails.deleteMany();
  const deleteProduct = prisma.product.deleteMany();
  const deleteCategory = prisma.category.deleteMany();
  const deleteCustomerOrder = prisma.customerOrder.deleteMany();
  const deleteCustomer = prisma.customer.deleteMany();

  await prisma.$transaction([deleteOrderDetails, deleteProduct, deleteCategory, deleteCustomerOrder, deleteCustomer]);

  await prisma.$disconnect();
});

it('should create 1 new customer with 1 order', async () => {
  // The new customers details
  const customer: Customer = {
    id: 2,
    name: 'Hermione Granger',
    email: 'hermione@hogwarts.io',
    address: '2 Hampstead Heath',
  };
  // The new orders details
  const order: OrderInput = {
    customer,
    productId: 1,
    quantity: 1,
  };

  // Create the order and customer
  await createOrder(order);

  // Check if the new customer was created by filtering on unique email field
  const newCustomer = await prisma.customer.findUnique({
    where: {
      email: customer.email,
    },
  });

  // Check if the new order was created by filtering on unique email field of the customer
  const newOrder = await prisma.customerOrder.findFirst({
    where: {
      customer: {
        email: customer.email,
      },
    },
  });

  // Expect the new customer to have been created and match the input
  expect(newCustomer).toEqual(customer);
  // Expect the new order to have been created and contain the new customer
  expect(newOrder).toHaveProperty('customerId', 2);
});

it('should create 1 order with an existing customer', async () => {
  // The existing customers email
  const customer: Customer = {
    email: 'harry@hogwarts.io',
  };
  // The new orders details
  const order: OrderInput = {
    customer,
    productId: 1,
    quantity: 1,
  };

  // Create the order and connect the existing customer
  await createOrder(order);

  // Check if the new order was created by filtering on unique email field of the customer
  const newOrder = await prisma.customerOrder.findFirst({
    where: {
      customer: {
        email: customer.email,
      },
    },
  });

  // Expect the new order to have been created and contain the existing customer with an id of 1 (Harry Potter from the seed script)
  expect(newOrder).toHaveProperty('customerId', 1);
});

it("should show 'Out of stock' message if productId doesn't exit", async () => {
  // The existing customers email
  const customer: Customer = {
    email: 'harry@hogwarts.io',
  };
  // The new orders details
  const order: OrderInput = {
    customer,
    productId: 3,
    quantity: 1,
  };

  // The productId supplied doesn't exit so the function should return an "Out of stock" message
  await expect(createOrder(order)).resolves.toEqual(new Error('Out of stock'));
});

*/
