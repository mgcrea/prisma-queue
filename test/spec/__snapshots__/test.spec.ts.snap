// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrismaQueue enqueue should properly enqueue a job 1`] = `
Array [
  "record",
]
`;

exports[`PrismaQueue should properly create a queue 1`] = `
Array [
  "_events",
  "_eventsCount",
  "_maxListeners",
  "options",
  "worker",
  "count",
  "concurrency",
  "stopped",
  "name",
  "config",
]
`;
